src/plone/restapi/indexers.py

```diff
@@ -55,59 +59,48 @@ def __call__(self, value):
         return result
 
 
+@implementer(IBlockSearchableText)
+@adapter(IBlocks, IBrowserRequest)
+class SlateTextIndexer:
+    """Searchable Text indexer for Slate blocks."""
+
+    def __init__(self, context, request):
+        self.context = context
+        self.request = request
+
+    def __call__(self, block):
+        block = block or {}
+        # BBB compatibility with slate blocks that used the "plaintext" field
+        return block.get("plaintext", "")
+
+
 @indexer(IBlocks)
 def SearchableText_blocks(obj):
+    """Extract text to be used by the SearchableText index in the Catalog."""
     request = getRequest()
     blocks = obj.blocks
+    blocks_layout = obj.blocks_layout
     blocks_text = []
-
-    for block in blocks.values():
-
+    for block_id in blocks_layout.get("items", []):
```

Comment:
I think we need to accept the reality of "blocks in blocks". I can mention already two scenarios that will surely affect you: the teasers grid (ok, maybe not the end of the world if that content is not searched) and the accordion block.

See https://github.com/plone/plone.restapi/blob/9c04a3c421300a5f7bfb6732cc0177fa302446a6/src/plone/restapi/deserializer/blocks.py#L54 which has support for deserializers.

I think we should have a separate "block traversing utility". If needed, I can provide some examples, but it would be the equivalent of this: https://github.com/plone/volto/blob/616c19f420aa3117ffd94b2dc99fe4c1da1f03bf/src/helpers/Blocks/Blocks.js#L348

Reply:
I do believe we should trust the "block container" to provide us with the **searchableText** attribute or force a standard way for blocks telling us they have sub-blocks (afaik, we do not have that yet as there are implementations with sub-blocks in blocks or data attributes).

Reply:
no... it's gonna be unwieldy to implement this Volto side. What do you do with a columns block inside an accordion block inside a tabs block? All of that extra weight in the json will start piling up, but it's really simple to handle this in the backend by following the convention on how block data would look like.

Reply:
> force a standard way for blocks telling us they have sub-blocks

we have a convention that's followed by the volto-columns-block, volto-accordion-block, volto-tabs-block (any many other EEA addons). The convention is `blocks+blocks_layout` either directly in the block data or in a `data` field of the block data. The volto-blocks-grid unfortunately doesn't follow the convention https://github.com/kitconcept/volto-blocks-grid/blob/da622c3b1635051be0d53476c784e379145db7de/src/components/Grid/Edit.jsx#L180 although I remember that at some point there was discussion to get that migrated, see Volto Team Meeting notes from 2021-10-12

Reply:
We might also do "typed values", see https://github.com/plone/volto/issues/2252

Reply:
So, we've actually had a initiative about a year ago to introduce a new interface function (something declared in a block's config) to allow a block to declare its searchableText, so that it can be lifted to its parent container block. But it wouldn't completely solve the problem, as the parent container block needs to know if it's inside another block. Or otherwise have an "onChange" callback built into the Edit and Add forms, which could recursively go through the blocks and call their api methods.

Reply:
I will add recursion here and follow the convention of "blocks" or "data" with "blocks", and search for searchableText attributes there.
Do you think this is a good compromise @tiberiuichim ?

Reply:
@ericof yes, that would be perfect.

Reply:
@tiberiuichim Implemented :-)